src/fs.ts

```diff
@@ -1,86 +1,90 @@
-import {
-	RequestType, IConnection
-} from 'vscode-languageserver';
+import { IConnection } from 'vscode-languageserver';
 
-import * as fs from 'fs';
-import * as path_ from 'path';
+import { TextDocumentIdentifier, TextDocumentItem } from 'vscode-languageserver-types';
 
-export interface FileInfo {
-	name: string
-	size: number
-	dir: boolean
-}
+import * as path from 'path';
+import * as fs from 'mz/fs';
+import glob = require('glob');
+import { path2uri, uri2path } from './util';
 
 export interface FileSystem {
-	readDir(path: string, callback: (err: Error, result?: FileInfo[]) => void): void
-	readFile(path: string, callback: (err: Error, result?: string) => void): void
-}
-
-export namespace ReadDirRequest {
-	export const type: RequestType<string, FileInfo[], any> = { get method() { return 'fs/readDir'; } };
-}
-
-export namespace ReadFileRequest {
-	export const type: RequestType<string, string, any> = { get method() { return 'fs/readFile'; } };
+	/**
+	 * Returns all files in the workspace under base
+	 *
+	 * @param base A URI under which to search, resolved relative to the rootUri
+	 * @return A promise that is fulfilled with an array of URIs
+	 */
+	getWorkspaceFiles(base?: string): Promise<string[]>;
+
+	/**
+	 * Returns the content of a text document
+	 *
+	 * @param uri The URI of the text document, resolved relative to the rootUri
+	 * @return A promise that is fulfilled with the text document content
+	 */
+	getTextDocumentContent(uri: string): Promise<string>;
 }
 
 export class RemoteFileSystem implements FileSystem {
 
 	private connection: IConnection;
+	private workspaceFilesPromise?: Promise<string[]>;
 
 	constructor(connection: IConnection) {
 		this.connection = connection
 	}
 
-	readDir(path: string, callback: (err: Error | null, result?: FileInfo[]) => void) {
-		this.connection.sendRequest(ReadDirRequest.type, path).then((f: FileInfo[]) => {
-			return callback(null, f)
-		}, callback)
+	/**
+	 * The files request is sent from the server to the client to request a list of all files in the workspace or inside the directory of the base parameter, if given.
+	 * A language server can use the result to index files by filtering and doing a content request for each text document of interest.
+	 */
+	getWorkspaceFiles(base?: string): Promise<string[]> {
+		// TODO cache this at a different layer and invalidate it properly
```

Comment:
What I don't like here is that we keep an array of the files here in memory, but `LocalFileSystem` essentially does too (with the difference that the former is never updated). I was thinking about a `WorkspaceTree` class that holds this state and optionally the content of files. Maybe also remove the dummy placeholder contents.